pinot-core/src/main/java/org/apache/pinot/core/operator/streaming/BaseStreamBlockCombineOperator.java

```diff
@@ -0,0 +1,176 @@
+    _numOperatorsFinished = 0;
+    _exceptionBlock = null;
+  }
+
+  @Override
+  protected BaseResultsBlock getNextBlock() {
+    long endTimeMs = _queryContext.getEndTimeMs();
+    while (_exceptionBlock == null && _numOperatorsFinished < _numOperators) {
+      try {
+        BaseResultsBlock resultsBlock =
+            _blockingQueue.poll(endTimeMs - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+        if (resultsBlock == null) {
+          // Query times out, skip streaming the remaining results blocks
+          LOGGER.error("Timed out while polling results block (query: {})", _queryContext);
+          _exceptionBlock = new ExceptionResultsBlock(QueryException.getException(
+              QueryException.EXECUTION_TIMEOUT_ERROR, new TimeoutException("Timed out while polling results block")));
+          return _exceptionBlock;
+        }
+        if (resultsBlock.getProcessingExceptions() != null) {
+          // Caught exception while processing segment, skip streaming the remaining results blocks and directly return
+          // the exception
+          _exceptionBlock = resultsBlock;
+          return _exceptionBlock;
+        }
+        if (resultsBlock == LAST_RESULTS_BLOCK) {
+          // Caught LAST_RESULTS_BLOCK from a specific task, indicated it has finished.
+          // Skip returning this metadata block and continue to process the next from the _blockingQueue.
+          _numOperatorsFinished++;
+          continue;
+        }
+        return resultsBlock;
+      } catch (Exception e) {
+        LOGGER.error("Caught exception while merging results blocks (query: {})", _queryContext, e);
+        _exceptionBlock = new ExceptionResultsBlock(QueryException.getException(QueryException.INTERNAL_ERROR, e));
+        return _exceptionBlock;
+      }
+    }
+    // Setting the execution stats for the final return
+    BaseResultsBlock finalBlock = new MetadataResultsBlock();
+    int numServerThreads = Math.min(_numTasks, ResourceManager.DEFAULT_QUERY_WORKER_THREADS);
+    CombineOperatorUtils.setExecutionStatistics(finalBlock, _operators, _totalWorkerThreadCpuTimeNs.get(),
+        numServerThreads);
+    return finalBlock;
+  }
+
+  /**
+   * Executes query on one or more segments in a worker thread.
+   */
+  @Override
+  protected void processSegments() {
+    int operatorId;
+    while ((operatorId = _nextOperatorId.getAndIncrement()) < _numOperators) {
+      Operator operator = _operators.get(operatorId);
+      T resultsBlock;
+      try {
+        if (operator instanceof AcquireReleaseColumnsSegmentOperator) {
+          ((AcquireReleaseColumnsSegmentOperator) operator).acquire();
+        }
+        while ((resultsBlock = (T) operator.nextBlock()) != null) {
+          if (shouldFinishStream(resultsBlock) || _combineFunction.isQuerySatisfied(resultsBlock)) {
+            // Query is satisfied, skip processing the remaining segments
+            _blockingQueue.offer(resultsBlock);
```

Comment:
When query is satisfied, we should skip processing all the following operators, instead of only breaking the current operator.

Reply:
correct. let me add a TODO and address in separate PR.

Reply:
discussed offline. we need to address in this PR in order to not introduce regression. will follow up before merging